lib/fluent/plugin/in_tail.rb

```diff
@@ -502,9 +502,13 @@ def update_watcher(tail_watcher, pe, new_inode)
       new_target_info = TargetInfo.new(path, new_inode)
 
       if @follow_inodes
-        # When follow_inodes is true, it's not cleaned up by refresh_watcher.
-        # So it should be unwatched here explicitly.
-        tail_watcher.unwatched = true
+        # If the old inode still exists, we can't unwatch it.
+        # If the old inode disappears, we should unwatch it here because
+        # `refresh_watcher` can't detect that the inode disappears.
+        # (It is because the path of `@tails` will be overwritten here.)
+        # (It is not a big problem if we don't unwatch here, as long as we make
+        # sure to detach it. The position will be cleared on restart eventually.)
+        tail_watcher.unwatched = true unless expand_paths.key?(tail_watcher.ino)
```

Comment:
I'm worry calling `expand_paths` here, because it's relatively high cost.
It's might to too high cost on heavy rotated environment.
In addition, the file might be disappear soon just after calling expand_paths, as described in https://github.com/fluent/fluentd/blob/691b2e0c39d7d93a7d648dc5308ff961a4d1bda1/lib/fluent/plugin/in_tail.rb#L337-L348
(Although it might not so big problem as you described in the comment).

Reply:
Thanks for your review!

> I'm worry calling expand_paths here, because it's relatively high cost.
> It might be too high cost on heavy rotated environment.

I see.
I thought it would be okay since it was only called on rotation, but certainly, this is worrisome considering the case of so many files being rotated so often...

It may be better to cache the list in `refresh_watcher` and use it here...
However, I'm unsure if I should go this far for unwatching here.

> (Although it might not so big problem as you described in the comment).

What we should be concerned about is that, without unwatching here, the memory and file for PositionFile would be larger.
(It will be cleared only when restarting. We can't use `pos_file_compaction_interval` to clear this because it doesn't check the target existence.)

Reply:
Possible fix:

* A: Don't unwatch here.
* B: Fix `refresh_watcher` to refresh PositionFile as well.
* C: Cache the list in `refresh_watcher` and use it in `update_watcher`
* D: Change the key of `@tails` to inode when enabling `follow_inodes`

Reply:
>     * A: Don't unwatch here.

We want to avoid this if possible since the memory and file for PositionFile would be larger.

>     * B: Fix `refresh_watcher` to refresh PositionFile as well.

It could be possible.

>     * C: Cache the list in `refresh_watcher` and use it in `update_watcher`

This is not good. It depends on the interval of `refresh_watcher`.

>     * D: Change the key of `@tails` to inode when enabling `follow_inodes`

It could be possible.
We should consider the following PR for this.

* #4185

Reply:
I close this PR since it should be fixed by B or D.
Are you OK for closing it?

Reply:
Yes! I close this! Thanks!